sdk/core/azure-core/inc/azure/core/http/winhttp/win_http_client.hpp

```diff
@@ -6,29 +6,103 @@
  */
 
 #pragma once
-#ifdef BUILD_TRANSPORT_WINHTTP_ADAPTER
+//#ifdef BUILD_TRANSPORT_WINHTTP_ADAPTER
 
 #include "azure/core/http/http.hpp"
 #include "azure/core/http/policy.hpp"
 
 #include <type_traits>
 #include <vector>
+#include <winhttp.h>
 
 namespace Azure { namespace Core { namespace Http {
 
+  namespace Details {
+
+    constexpr static int64_t DefaultUploadChunkSize = 1024 * 64;
+    constexpr static int64_t MaximumUploadChunkSize = 1024 * 1024 * 1024;
+
+    class WinHttpStream : public BodyStream {
+    private:
+      HINTERNET m_sessionHandle;
+      HINTERNET m_connectionHandle;
+      HINTERNET m_requestHandle;
+      bool m_isEOF;
```

Comment:
That won't work if content length is -1. Not sure if there is much value in adding a function like this, versus just setting the flag, especially since the m_iEOF is set within the loop atm.

Reply:
You know, reducing the size if the state we maintain is always better.
The more state we have, the bigger the chances for bugs. The problem starts when the state is changed from a new place and then another, and another.

So, as a good practice, if any state is defined by some other state (like in this case), we can avoid maintaining this new state by just having a function that describes it (stateless programming is beautiful :D).

For the case of length is -1, we would only need to update the isOOF() function to add this condition and define what does EOF means in this case (instead of maintaining all the -1 lenght program flow to make check what's the value that is set to m_isEOF

Reply:
How would the check for `numberOfBytesRead == 0` work then? We want to set IsEOF to be true when a call to `WinHttpReadData` returns 0 bytes read.

If content length is -1, we are supposed to keep reading until there is no more data. How would that be encapsulated in the `IsEOF` function without additional state tracking?

Currently, this is the condition for when I set the `m_isEOF` state:
```C++
if (numberOfBytesRead == 0
      || (this->m_contentLength != -1
          && this->m_streamTotalRead == this->m_contentLength))
  {
    this->m_isEOF = true;
  }
```

 Are you suggesting I don't check that and only have this?
```C++
bool IsEOF()
{
  int64_t length = Length();
  return length != -1 && this->m_streamTotalRead == length;
}
```

Reply:
Based on offline conversation, leaving it as is, since this is an implementation detail choice and functionally correct.